CMakeLists.txt

```diff
@@ -40,36 +49,31 @@ else()
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops")
 endif()
 
+# Convenience routine to apply list of target_compile_definitions()
+set(xgboost_defs "")
+
 # AVX
 if(USE_AVX)
   if(MSVC)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /arch:AVX")
   else()
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mavx")
   endif()
-  add_definitions(-DXGBOOST_USE_AVX)
+  list(APPEND xgboost_defs XGBOOST_USE_AVX)
 endif()
 
-
 # compiled code customizations for R package
 if(R_LIB)
-  add_definitions(
-    -DXGBOOST_STRICT_R_MODE=1
-    -DXGBOOST_CUSTOMIZE_GLOBAL_PRNG=1
-    -DDMLC_LOG_BEFORE_THROW=0
-    -DDMLC_DISABLE_STDIN=1
-    -DDMLC_LOG_CUSTOMIZE=1
-    -DRABIT_CUSTOMIZE_MSG_
-    -DRABIT_STRICT_CXX98_
-  )
+  list(APPEND xgboost_defs
+    XGBOOST_STRICT_R_MODE=1
+    XGBOOST_CUSTOMIZE_GLOBAL_PRNG=1
+    DMLC_LOG_BEFORE_THROW=0  # DMLC: should be passed in config
```

Comment:
These flags seem to be preferred for the R package.  I wouldn't use this build mode, but it is quite popular in the data science community (many xgboost users).  Prior to hunter + find_package() configurations the `rabbit` and `dmlc-core` libs were built using `add_subdirectory()` so these flags were globally defined and would be applied properly.  In the package management case, I suppose we would have to introduce CMake options in `dmlc-core` and `rabit` to allow setting these from a top level `config.cmake` as needed.  I did test the `R` build.

Reply:
> I suppose we would have to introduce CMake options in dmlc-core and rabit to allow setting these from a top level config.cmake as needed

Yep